File: backend/src/main/java/com/heartcare/patient/model/PatientProfile.java
package com.heartcare.patient.model;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.Id;
import jakarta.persistence.PrePersist;
import jakarta.persistence.PreUpdate;
import jakarta.persistence.Table;
import org.hibernate.annotations.JdbcTypeCode;
import org.hibernate.type.SqlTypes;

import java.time.OffsetDateTime;
import java.util.ArrayList;
import java.util.List;
import java.util.UUID;

@Entity
@Table(name = "patient_profiles")
public class PatientProfile {

    @Id
    @Column(name = "user_id")
    private UUID userId;

    @Column(name = "birth_year")
    private Integer birthYear;

    @Column(name = "preferred_language", length = 5)
    private String preferredLanguage;

    @Column(name = "height_cm")
    private Integer heightCm;

    @Column(name = "chd_stage", length = 50)
    private String chdStage;

    @Column(name = "disease_history")
    private String diseaseHistory;

    @JdbcTypeCode(SqlTypes.JSON)
    @Column(name = "comorbidities", nullable = false)
    private List<String> comorbidities = new ArrayList<>();

    @Column(name = "management_plan")
    private String managementPlan;

    @JdbcTypeCode(SqlTypes.JSON)
    @Column(name = "goals")
    private Goals goals;

    @Column(name = "created_at", nullable = false)
    private OffsetDateTime createdAt;

    @Column(name = "updated_at", nullable = false)
    private OffsetDateTime updatedAt;

    protected PatientProfile() {
        // for JPA
    }

    public PatientProfile(UUID userId) {
        this.userId = userId;
    }

    @PrePersist
    void onCreate() {
        OffsetDateTime now = OffsetDateTime.now();
        if (createdAt == null) {
            createdAt = now;
        }
        updatedAt = now;
    }

    @PreUpdate
    void onUpdate() {
        updatedAt = OffsetDateTime.now();
    }

    public UUID getUserId() {
        return userId;
    }

    public Integer getBirthYear() {
        return birthYear;
    }

    public void setBirthYear(Integer birthYear) {
        this.birthYear = birthYear;
    }

    public String getPreferredLanguage() {
        return preferredLanguage;
    }

    public void setPreferredLanguage(String preferredLanguage) {
        this.preferredLanguage = preferredLanguage;
    }

    public Integer getHeightCm() {
        return heightCm;
    }

    public void setHeightCm(Integer heightCm) {
        this.heightCm = heightCm;
    }

    public String getChdStage() {
        return chdStage;
    }

    public void setChdStage(String chdStage) {
        this.chdStage = chdStage;
    }

    public String getDiseaseHistory() {
        return diseaseHistory;
    }

    public void setDiseaseHistory(String diseaseHistory) {
        this.diseaseHistory = diseaseHistory;
    }

    public List<String> getComorbidities() {
        return comorbidities;
    }

    public void setComorbidities(List<String> comorbidities) {
        this.comorbidities = (comorbidities == null) ? new ArrayList<>() : comorbidities;
    }

    public String getManagementPlan() {
        return managementPlan;
    }

    public void setManagementPlan(String managementPlan) {
        this.managementPlan = managementPlan;
    }

    public Goals getGoals() {
        return goals;
    }

    public void setGoals(Goals goals) {
        this.goals = goals;
    }

    public OffsetDateTime getCreatedAt() {
        return createdAt;
    }

    public OffsetDateTime getUpdatedAt() {
        return updatedAt;
    }
}

File: mobile/lib/features/profile/domain/entities/patient_profile.dart
import 'health_goals.dart';

class PatientProfile {
  const PatientProfile({
    required this.userId,
    required this.birthYear,
    required this.preferredLanguage,
    required this.heightCm,
    required this.chdStage,
    required this.diseaseHistory,
    required this.comorbidities,
    required this.managementPlan,
    required this.goals,
    required this.updatedAt,
  });

  factory PatientProfile.empty(String userId) => PatientProfile(
    userId: userId,
    birthYear: null,
    preferredLanguage: null,
    heightCm: null,
    chdStage: null,
    diseaseHistory: null,
    comorbidities: const <String>[],
    managementPlan: null,
    goals: null,
    updatedAt: null,
  );

  final String userId;
  final int? birthYear;
  final String? preferredLanguage;
  final double? heightCm;
  final String? chdStage;
  final String? diseaseHistory;
  final List<String> comorbidities;
  final String? managementPlan;
  final HealthGoals? goals;
  final DateTime? updatedAt;

  PatientProfile copyWith({
    int? birthYear,
    String? preferredLanguage,
    double? heightCm,
    String? chdStage,
    String? diseaseHistory,
    List<String>? comorbidities,
    String? managementPlan,
    HealthGoals? goals,
    DateTime? updatedAt,
  }) {
    return PatientProfile(
      userId: userId,
      birthYear: birthYear ?? this.birthYear,
      preferredLanguage: preferredLanguage ?? this.preferredLanguage,
      heightCm: heightCm ?? this.heightCm,
      chdStage: chdStage ?? this.chdStage,
      diseaseHistory: diseaseHistory ?? this.diseaseHistory,
      comorbidities: comorbidities ?? this.comorbidities,
      managementPlan: managementPlan ?? this.managementPlan,
      goals: goals ?? this.goals,
      updatedAt: updatedAt ?? this.updatedAt,
    );
  }

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is PatientProfile &&
          runtimeType == other.runtimeType &&
          userId == other.userId &&
          birthYear == other.birthYear &&
          preferredLanguage == other.preferredLanguage &&
          heightCm == other.heightCm &&
          chdStage == other.chdStage &&
          diseaseHistory == other.diseaseHistory &&
          _listEquals(comorbidities, other.comorbidities) &&
          managementPlan == other.managementPlan &&
          goals == other.goals &&
          updatedAt == other.updatedAt;

  @override
  int get hashCode => Object.hash(
    userId,
    birthYear,
    preferredLanguage,
    heightCm,
    chdStage,
    diseaseHistory,
    Object.hashAll(comorbidities),
    managementPlan,
    goals,
    updatedAt,
  );
}

bool _listEquals(List<String> a, List<String> b) {
  if (a.length != b.length) return false;
  for (var i = 0; i < a.length; i++) {
    if (a[i] != b[i]) return false;
  }
  return true;
}
